src/libstd/collections/hash/map.rs

```diff
@@ -2445,13 +2445,18 @@ impl RandomState {
         // iteration order allows a form of DOS attack. To counter that we
         // increment one of the seeds on every RandomState creation, giving
         // every corresponding HashMap a different iteration order.
-        thread_local!(static KEYS: Cell<(u64, u64)> = {
-            Cell::new(sys::hashmap_random_keys())
+        thread_local!(static KEYS: Cell<[u64; 2]> = {
+            let mut buf = [0u8; 16];
+            // we consciously ignore potential errors here
+            let _ = getrandom::getrandom(&mut buf);
+            // SAFETY: it's safe to transmute_copy random 16 bytes to two `u64`s
+            let buf: [u64; 2] = unsafe { core::mem::transmute_copy(&buf) };
```

Comment:
`transmute_copy` does not do the check that the sizes has to be the same, right? In that case I think using `transmute` is better as it checks more things.

There was no problem with the previous code.

Reply:
Hm, I thought the following code is incorrect, due to the broken alignment. Am I wrong?
```rust
#[repr(packed)]
struct Foo {
    a: u8,
    b: [u8; 64]
}

fn foo(val: Foo) -> [u64; 8] {
    unsafe {
        std::mem::transmute(val.b)
    }
}
```

Yes, `transmute_copy` does not check sizes, but I think it shouldn't be a problem as we can easily see sizes of input and output and they are defined locally.

Reply:
Yes you are wrong.

What is incorrect is to turn an `&[u8; 64]` into an `&[u64; 8]`. *References* and *memory accesses* must be sufficiently aligned:
```rust
fn main() { unsafe {
    let x = &[0u8; 64];
    let y = &*(x as *const _ as *const [u64; 8]); // UB: Creating an unaligned references
    let _val = *(x as *const _ as *const [u64; 8]); // UB: Performing an unaligned load
} }
```

But when copying *values* around, how would alignment even enter the equation? You have to make sure that every time a load or store is performed, the alignment is satisfied, and we also mandate that references are sufficiently aligned at any time. Which load or store or reference in your program violates alignment?

You can also [use Miri to test this](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=9955cca88cc06eed7339926f48c5f709) (Miri is in the "Tools" menu).